arangod/Aql/Condition.cpp

```diff
@@ -1035,6 +1036,99 @@ void Condition::collectOverlappingMembers(
+      otherArrayNode = otherArrayNode->getMember(0);
+
+      TRI_ASSERT(arrayNode->numMembers() == 5);
+      if (arrayNode->numMembers() != otherArrayNode->numMembers()) {
+        continue;
+      }
+
+      // Check if the checked element from IN_RANGE function is part of the
+      // index
+      auto const* firstElemInRangeFunction = arrayNode->getMember(0);
+      if (firstElemInRangeFunction->type != NODE_TYPE_ATTRIBUTE_ACCESS ||
+          !firstElemInRangeFunction->isAttributeAccessForVariable(
+              result, isFromTraverser) ||
+          result.first != variable ||
+          !basics::AttributeName::isIdentical(result.second, index->fields()[0],
+                                              false)) {
+        ::clearAttributeAccess(result);
+        continue;
+      }
+      ::clearAttributeAccess(result);
+
+      bool safeToRemove = true;
+      for (size_t i = 0; i < arrayNode->numMembers() && safeToRemove; ++i) {
+        auto const* functionCallNodeElement = arrayNode->getMemberUnchecked(i);
+        auto const* functionCallNodeOtherElement =
+            otherArrayNode->getMemberUnchecked(i);
+
+        if (functionCallNodeElement->type !=
+            functionCallNodeOtherElement->type) {
+          safeToRemove = false;
+          break;
+        }
+
+        switch (functionCallNodeElement->type) {
+          case NODE_TYPE_ATTRIBUTE_ACCESS: {
+            std::pair<Variable const*, std::vector<basics::AttributeName>>
+                attributeAccessForVariableResult;
+            std::pair<Variable const*, std::vector<basics::AttributeName>>
+                attributeAccessForVariableOtherResult;
+
+            if (!functionCallNodeElement->isAttributeAccessForVariable(
+                    attributeAccessForVariableResult, isFromTraverser) ||
+                !functionCallNodeOtherElement->isAttributeAccessForVariable(
+                    attributeAccessForVariableOtherResult, isFromTraverser) ||
+                attributeAccessForVariableResult.first !=
+                    attributeAccessForVariableOtherResult.first ||
+                !basics::AttributeName::isIdentical(
+                    attributeAccessForVariableResult.second,
+                    attributeAccessForVariableOtherResult.second, false)) {
+              safeToRemove = false;
+            }
+            break;
+          }
+          case NODE_TYPE_VALUE: {
+            if (aql::compareAstNodes(functionCallNodeElement,
+                                     functionCallNodeOtherElement, true) != 0) {
+              safeToRemove = false;
+            }
+            break;
+          }
+          default: {
+            // We are being extremely pessimistic, meaning if we encounter
+            // anything else we will not remove the expression
+            safeToRemove = false;
+            break;
+          }
+        }
+      }
+
+      if (safeToRemove) {
+        toRemove.emplace(i);
+      }
+    }
```

Comment:
Would it make sense to extract this into a separate function? In particular because the upcoming vector index PR maybe adds another case to this function.